### DISCUSSION TITLE
Application with multiple windows

### DISCUSSION THREAD

**Discussion by akhilman** [Relm4/Relm4#80] [Q&A]
Is there a way to make application with multiple similar windows?
Can someone provide an example?

**Comment by edfloreshz** [comment]
Hi, sorry this took so long.

You need to create a new component that has `gtk::Window` or `adw::Window` as its root widget.

Then, store that component's model inside your parent component:

```rust
struct ParentModel {
    child: Controller<ChildModel>
}
```

After that, you need to trigger an event to open the child window, inside that event, you can get a reference to the widget:

`self.child.widget()`

And then call `present()`

```rust
ParentEvent::OpenChildWindow => {
    let child_window = self.child.widget();
    child_window.present()
}
```

You can optionally add an `adw::HeaderBar` to allow closing the child window.

**Reply by tiago-vargas** [reply]
This approach works fine as long as you don't try to reopen the child window after it has been closed, since you call `present()` always on the same child.

Creating a new window every time avoids this:

``` rs
struct ParentModel {
    // child: Controller<ChildModel>
}
```

``` rs
ParentEvent::OpenChildWindow => {
    let child_window = ChildModel::builder()
        .launch(ChildInit)
        .detach();
    child_window.widget().present()
}
```

To set the child window transient for the parent, you can query the parent window:

``` rs
ParentEvent::OpenChildWindow => {
    let app = relm4::main_application();
    let parent_window = app.windows().first()
        .expect("Event should have been triggered from last focused window, thus first item")
        .clone();

    let child_window = ChildModel::builder()
        .transient_for(&parent_window)
        .launch(ChildInit)
        .detach();

    child_window.widget().present()
}
```

I made the assumption that `OpenChildWindow` is always triggered from the parent window.

This worked well when I was trying to open a preferences window from a menu in the main window's header bar (after also applying `set_modal: true` on the child window).
To open the menu, the main window would naturally be focused and it would be the first item in `windows()`.
